lib/babe/helpers_test.go

```diff
@@ -133,12 +159,209 @@ func NewTestService(t *testing.T, cfg *core.Config) *core.Service {
+			Keystore:             rtCfg.Keystore,
+			Network:              mockNetwork,
+			CodeSubstitutedState: dbSrv.Base,
+			CodeSubstitutes:      make(map[common.Hash]string),
+		}
+
+		babeService.blockImportHandler = NewTestService(t, &coreConfig, genesis,
+			genesisTrie, genesisHeader)
+	}
+
+	return babeService
+}
+
+func newTestServiceSetupParameters(t *testing.T, genesis genesis.Genesis,
+	genesisTrie trie.Trie, genesisHeader types.Header) (*Service, *state.EpochState, *types.BabeConfiguration) {
+	ctrl := gomock.NewController(t)
+	telemetryMock := NewMockTelemetry(ctrl)
+	telemetryMock.EXPECT().SendMessage(gomock.Any()).AnyTimes()
+
+	testDatadirPath := t.TempDir()
+
+	config := state.Config{
+		Path:      testDatadirPath,
+		LogLevel:  log.Info,
+		Telemetry: telemetryMock,
+	}
+	dbSrv := state.NewService(config)
+	dbSrv.UseMemDB()
+
+	err := dbSrv.Initialise(&genesis, &genesisHeader, &genesisTrie)
+	require.NoError(t, err)
+
+	err = dbSrv.Start()
+	require.NoError(t, err)
+
+	t.Cleanup(func() {
+		_ = dbSrv.Stop()
+	})
+
+	rtCfg := wasmer.Config{
+		Storage: rtstorage.NewTrieState(&genesisTrie),
+	}
+
+	rt, err := wasmer.NewRuntimeFromGenesis(rtCfg)
+	require.NoError(t, err)
+
+	genCfg, err := rt.BabeConfiguration()
+	require.NoError(t, err)
+
+	s := &Service{
+		epochState: dbSrv.Epoch,
+	}
+
+	return s, dbSrv.Epoch, genCfg
+}
+
+func createSecondaryVRFPreDigest(t *testing.T,
+	keypair *sr25519.Keypair, index uint32,
+	slot, epoch uint64, randomness Randomness,
+) *types.BabeSecondaryVRFPreDigest {
+	transcript := makeTranscript(randomness, slot, epoch)
+	out, proof, err := keypair.VrfSign(transcript)
+	require.NoError(t, err)
+
+	return types.NewBabeSecondaryVRFPreDigest(index, slot, out, proof)
+}
+
+func buildLocalTransaction(t *testing.T, rt runtime.Instance, ext types.Extrinsic,
+	bestBlockHash common.Hash) types.Extrinsic {
+	runtimeVersion := rt.Version()
+	txQueueVersion, err := runtimeVersion.TaggedTransactionQueueVersion()
+	require.NoError(t, err)
+	var extrinsicParts [][]byte
+	switch txQueueVersion {
+	case 3:
+		extrinsicParts = [][]byte{{byte(types.TxnLocal)}, ext, bestBlockHash.ToBytes()}
+	case 2:
+		extrinsicParts = [][]byte{{byte(types.TxnLocal)}, ext}
+	}
+	return types.Extrinsic(bytes.Join(extrinsicParts, nil))
+}
+
+func getSlot(t *testing.T, rt runtime.Instance, timestamp time.Time) Slot {
+	t.Helper()
+	babeConfig, err := rt.BabeConfiguration()
+	require.NoError(t, err)
+
+	currentSlot := uint64(timestamp.UnixMilli()) / babeConfig.SlotDuration
+	return Slot{
+		start:    timestamp,
+		duration: time.Duration(babeConfig.SlotDuration) * time.Millisecond,
+		number:   currentSlot,
+	}
+}
+
+func createTestBlockWithSlot(t *testing.T, babeService *Service, parent *types.Header,
+	exts [][]byte, epoch uint64, epochData *epochData, slot Slot) *types.Block {
+	for _, ext := range exts {
+		vtx := transaction.NewValidTransaction(ext, &transaction.Validity{})
+		_, err := babeService.transactionState.Push(vtx)
```

Comment:
nit `vtx` -> `validTransaction`

```suggestion
		validTransaction := transaction.NewValidTransaction(ext, &transaction.Validity{})
		_, err := babeService.transactionState.Push(validTransaction)
```